images/virtualization-artifact/pkg/controller/vmop/internal/protection.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2024 Flant JSC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package internal
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/deckhouse/virtualization-controller/pkg/controller/vmop/internal/state"
+	virtv2 "github.com/deckhouse/virtualization/api/core/v1alpha2"
+)
+
+const protectionHandlerName = "ProtectionHandler"
+
+// ProtectionHandler manages finalizers on VirtualMachineOperation resource.
+type ProtectionHandler struct {
+	logger *slog.Logger
+	client client.Client
+}
+
+func NewProtectionHandler(logger *slog.Logger, client client.Client) *ProtectionHandler {
+	return &ProtectionHandler{
+		logger: logger.With("handler", protectionHandlerName),
+		client: client,
+	}
+}
+
+func (h ProtectionHandler) Handle(ctx context.Context, s state.VMOperationState) (reconcile.Result, error) {
+	if s.VirtualMachineOperation() == nil {
+		return reconcile.Result{}, nil
+	}
+
+	changed := s.VirtualMachineOperation().Changed()
+	log := h.logger.With("name", changed.GetName(), "namespace", changed.GetNamespace())
+
+	vm, _ := s.VirtualMachine(ctx)
+
+	// The only case when we need finalizer on VirtualMachineOperation is when operation is in progress.
+	if changed.DeletionTimestamp == nil && changed.Status.Phase == virtv2.VMOPPhaseInProgress {
+		log.Debug("Add protection finalizer for the InProgress phase")
+		controllerutil.AddFinalizer(changed, virtv2.FinalizerVMOPCleanup)
```

Comment:
We lost the distinction between cleanup and protection finalizers. Are you sure we need cleanup here? It looks like we need protection.

Reply:
Renamed to deletion. Do not add protection finalizers on VMs any more.